### DISCUSSION TITLE
How to manage Podman as a non-root user like with Docker?

### DISCUSSION THREAD

**Discussion by piotr-dobrogost** [containers/podman#24074] [General]
I've been using PyCharm for developing and debugging Python application in container with docker configured to be managed as a non-root user using steps shown at https://docs.docker.com/engine/install/linux-postinstall/#manage-docker-as-a-non-root-user
I wanted to move from docker to podman but can't find analogues information for podman. Changing docker to podman results in PyCharm not being able to communicate with podman using podman's unix socket as it's only accessible by root. 
I would like to leave the whole subject of using rootless podman out of discussion as it brings its own set of problems (mainly around network configuration).
Here is (sadly minimal) documentation on using PyCharm with podman – https://www.jetbrains.com/help/pycharm/podman.html

**Comment by rhatdan** [comment]
If you want to leak podman.socket into container you need to disable SELinux separation, since this is a very dangerous thing from a security point of view.  You would need to do same thing with Docker, if it did not have SELinux disabled by default.

podman run --security-opt label=disable ...

**Reply by piotr-dobrogost** [reply]
I don't know what you mean by _leak podman.socket into container_. I just want to grant my normal user on host (no container) access to `/run/podman/podman.sock` the same way I granted access to `/var/run/docker.sock` when using Docker. When using Docker I added my user to `docker` group which was enough to access `/var/run/docker.sock` because this group has access to this socket:

```
[piotr@demon]~% ll /var/run/docker.sock
srw-rw---- 1 root docker 0 wrz 26 14:22 /var/run/docker.sock
```
As `/run/podman/podman.sock`'s group is `root` I added my user to `root` group but it's not enough as Podman's socket is placed inside `/run/podman` directory with no group permissions. So it's necessary to grant additional permissions for group – `sudo chmod g+rwx /run/podman`. After this my IDE (PyCharm) is able to access `/run/podman/podman.sock` (through `/var/run/docker.sock` symlink).
Now the problem is that `/run/podman` is recreated after each reboot and the additional permissions are lost.
Is there a way to specify `/run/podman`'s permissions somehow in Podman's configuration?
Btw, is there any reason that `/run/podman` directory does not have `rw` access rights for its `root` group (making all these easier)?

As to SELinux I have it set in permissive mode so I think it doesn't block anything.

**Reply by piotr-dobrogost** [reply]
It seems https://github.com/containers/podman/discussions/24103 asks about the same thing and discovers the same reason for not being able to access Podman's socket as another user.
Can somebody please tell me why there is no information on this whereas for Docker it's part of the docs at https://docs.docker.com/engine/install/linux-postinstall/#manage-docker-as-a-non-root-user ?

**Reply by afbjorklund** [reply]
You are supposed to use "sudo" (or connect as root), in order to access the /run/podman/podman.sock.
Podman does not have a "podman" group, since users are supposed to run their own rootless containers.

You can find the details (systemd-tmpfiles) in this issue:
* https://github.com/containers/podman/issues/6809

**Reply by afbjorklund** [reply]
Podman Desktop has the same issue, but it doesn't have any "sudo" workaround built-in:

* https://github.com/containers/podman-desktop/issues/2861

**Reply by piotr-dobrogost** [reply]
@afbjorklund Thanks for your answers and linking to other issues.

> Podman does not have a "podman" group, since users are supposed to run their own rootless containers.

I know _users are supposed to run their own rootless containers_ but as I wrote in my question:

> I would like to leave the whole subject of using rootless podman out of discussion as it brings its own set of problems (mainly around network configuration)

which is basically the same as your reservation in https://github.com/containers/podman/issues/6809:

> Running rootless isn't the question here, it's about root.

As to
> You are supposed to use "sudo" (or connect as root), in order to access the /run/podman/podman.sock.

Well, you can use `sudo` when executing commands. How can you use `sudo` to access socket file? I think PyCharm communicates with Docker/Podman mainly using socket file which you set in the settings although there is a place to set Docker's binary in the settings, too.

**Reply by afbjorklund** [reply]
That was the original question, and minikube ended up adding a "podman" group to make it work the same way as "docker":

https://github.com/containers/podman/issues/6809#issuecomment-662031057

When talking to Podman Machine (VM), podman connects as root@ with ssh - but this access was blocked in the minikube VM

i.e. `podman machine init --rootful`

**Comment by afbjorklund** [comment]
> I would like to leave the whole subject of using rootless podman out of discussion as it brings its own set of problems

I'm not sure that you _can_ leave it out of the discussion, since it is the root of your problems - the "old way" is unsupported.

---

For Docker, running as root (with sudo) is the default - setting up the docker group is optional, setting up rootless is optional:

* https://docs.docker.com/engine/install/linux-postinstall/#manage-docker-as-a-non-root-user

> The Docker daemon binds to a Unix socket, not a TCP port. [Note: you _can_ [bind it to legacy port 2376](https://docs.docker.com/engine/security/protect-access/#use-tls-https-to-protect-the-docker-daemon-socket)]
> By default it's the `root` user that owns the Unix socket, and other users can only access it using `sudo`. 

* https://docs.docker.com/engine/security/rootless/ `dockerd-rootless-setuptool.sh install`

> [INFO] Make sure the following environment variables are set (or add them to ~/.bashrc):
>
> export PATH=/usr/bin:$PATH
> export DOCKER_HOST=unix:///run/user/1000/docker.sock

For Podman, it is the daemon and the socket that is optional - the default is to run daemonless, at least when running locally.

**Comment by rhatdan** [comment]
I have been writing about how bad an idea this is going back to 2015.

https://projectatomic.io/blog/2015/08/why-we-dont-let-non-root-users-run-docker-in-centos-fedora-or-rhel/
https://danwalsh.livejournal.com/78373.html

I believe this is the most dangerous thing you can do on a Linux System, more dangerous then setting up sudo without password, so podman should not encourage it when there are better ways to handle it.

**Reply by afbjorklund** [reply]
Even docker will not do it for you, even if they do show you how to do it (i.e. load the gun and help you aim it at foot...)

**Reply by afbjorklund** [reply]
For the minikube use case, it was running in a dedicated virtual machine (and with passwordless sudo already set up)
